Updated for GNOME 49

mendres82 opened the pull request:
Updated to reflect GNOME 49 API changes:

Meta.Window
Meta.Window removes the Meta.MaximizeFlags from the parameters for maximize() and unmaximize(). Also get_maximized() is removed and you can use is_maximized() instead. To maximize or unmaximize the window horizontally or vertically, you can call new methods set_maximize_flags() and set_unmaximize_flags().

mendres82 committed:
Updated for GNOME 49

gadicc commented:
I wonder if we could still support older versions with something like:

```diff
-            const currentMetaMaximized = metaWindow.get_maximized();
+            const currentMetaMaximized = (metaWindow.get_maximized || metaWindow.is_maximized)();
```

mendres82 commented:
If we had to maintain compatibility to older GNOME versions, I'd probably write separate blocks, because

1. it handles both changed methods metaWindow.is_maximized() and metaWindow.maximize() correctly and
2. makes the code easy to maintain, once you decide to remove the support for a specific version.
3. The code is clear in what it's doing.

```
const Config = imports.misc.config;
const shellVersion = parseFloat(Config.PACKAGE_VERSION);

if (shellVersion < 49) {
    const currentMetaMaximized = metaWindow.get_maximized();
    if (currentMetaMaximized !== Meta.MaximizeFlags.BOTH) {
        this._log.debug(`Maximizing ${metaWindow.get_title()}`);
        metaWindow.maximize(savedMetaMaximized);
    }
} else {
    const currentMetaMaximized = metaWindow.is_maximized();
    if (currentMetaMaximized !== Meta.MaximizeFlags.BOTH) {
        this._log.debug(`Maximizing ${metaWindow.get_title()}`);
        metaWindow.maximize();
    }                
}
```
Please note: I didn't test any of this yet, it's just a suggestion.

gadicc commented:
Yeah good point.  What's your preference, @nlpsuge?  Should we drop old versions or still support them for a while?

mendres82 committed:
Added zip-archive

BashBandito commented:
When can we expect this to hit the GNOME repo?

By the way, appreciate the work you all do. Love this extension!

mendres82 commented:
As long as this PR doesn't get merged or the extension gets updated by the author, you shouldn't expect it to hit GNOME extension website anytime soon. I don't want to put a forked version anywhere officially, because the author deserves all credit for this extension.

BashBandito commented:
@mendres82 

Oh, I see, it sounds like the author is absent. I see their last activity on their profile was in August, so perhaps they are busy with schooling?

That's understandable, but also rather unfortunate because this extension is one that brings a ton of value to daily workflow. Pretty rough without it when you have dual monitors, and you have to constantly fix your window positioning.

Do you happen to know if it works when changing the metadata.json version number? I've never gone that route before, but this one has me curious, as I am rather impatient for it.

mendres82 commented:
There is a [zip-file](https://github.com/nlpsuge/gnome-shell-extension-another-window-session-manager/raw/df587e0630bef738d881b1aa6986b46c207c1d6d/another-window-session-manager@gmail.com.zip) attached to my PR, which you can use to install the updated version while waiting for an official update.

BashBandito commented:
@mendres82 

Thank you for sharing this. I installed it, logged out and back in and there she is again.

You made my day.

magicman3211 commented:
without sounding too noobish, how can I install from zip file mentioned above @mendres82

BashBandito commented:
> without sounding too noobish, how can I install from zip file mentioned above @mendres82

Unzip the folder into `/.local/share/gnome-shell/extensions` or your equivalent

mendres82 commented:
Extensions can either be local or system-wide. Local extensions are ones installed by each user in their home directories (`~/.local/share/gnome-shell/extensions`), whereas system-wide extensions are installed by administrators in system directories (`/usr/share/gnome-shell/extensions/`). System extensions, therefore, cannot be installed, updated or removed by non-administrator users. They can, however, be enabled or disabled by each user.

[Source](https://docs.fedoraproject.org/en-US/quick-docs/gnome-shell-extensions/)

nlpsuge commented:
> What's your preference, @nlpsuge? Should we drop old versions or still support them for a while?

Hi, bonsoir, sorry for the late reply. 

(My phone was damaged for about 2 months, so I couldn't access my github, I just got it fixed at the end of December.)

Yes, I prefer to support the old versions. 


> If we had to maintain compatibility to older GNOME versions, I'd probably write separate blocks, because
> 
> 1. it handles both changed methods metaWindow.is_maximized() and metaWindow.maximize() correctly and
> 2. makes the code easy to maintain, once you decide to remove the support for a specific version.
> 3. The code is clear in what it's doing.
> 
> ```
> const Config = imports.misc.config;
> const shellVersion = parseFloat(Config.PACKAGE_VERSION);
> 
> if (shellVersion < 49) {
>     const currentMetaMaximized = metaWindow.get_maximized();
>     if (currentMetaMaximized !== Meta.MaximizeFlags.BOTH) {
>         this._log.debug(`Maximizing ${metaWindow.get_title()}`);
>         metaWindow.maximize(savedMetaMaximized);
>     }
> } else {
>     const currentMetaMaximized = metaWindow.is_maximized();
>     if (currentMetaMaximized !== Meta.MaximizeFlags.BOTH) {
>         this._log.debug(`Maximizing ${metaWindow.get_title()}`);
>         metaWindow.maximize();
>     }                
> }
> ```
> 
> Please note: I didn't test any of this yet, it's just a suggestion.


Good evening, @mendres82 , could you please commit this code and test it? I will test it later.

mendres82 commented:
Currently I'm not working too much on my private PC, so it might take a few days to get the fix ready and properly test it.

nlpsuge commented:
> so perhaps they are busy with schooling?

Hi, good morning / bonjour. That’s actually a good guess. I’m studying in France as an international student, and I’m focusing on learning French before enrolling in a master’s program. With learning French and going through the visa application process, it’s been a super busy and stressful year.

mendres82 committed:
Updated for GNOME 49, restored compatibility to old versions

mendres82 committed:
Removed zip-archive

mendres82 commented:
>> Good evening, @mendres82 , could you please commit this code and test it? I will test it later.

- Added compatibility for GNOME 49
- Restored compatibility to older versions
- Fixed bug in my old implementation, now correct handling of maximized window status

I tested window saving / restoring functionality on both openSUSE Tumbleweed (GNOME 49.2) and Ubuntu 24.04 LTS (GNOME 46), so it _should_ be fine.

gadicc commented:
Hey, @nlpsuge.  I'm no stranger to visas and language learning so welcome back and good luck with everything!! :D

@mendres82, thanks so much again for the original patch and for this latest commit so quickly.  Really appreciate that.

I'll just also add that I've been using the original patch successfully for about 2.5 months now on ArchLinux with GNOME 49 (I think since 49.0 and now 49.2).  Can't live without it! :D

mendres82 committed:
Improved code readability

mendres82 committed:
Fixed restoring unmaximized windows when currently maximized

mendres82 commented:
@nlpsuge: After testing some more, I found an issue in my code when restoring unmaximized windows that were currently maximized, so that should be working as intended now on GNOME 49+ again. The handling of maximized windows was broken in my first implementation, that's why I bumped the version to 52 instead of 51, so that users who use my unofficial zip get the update as well.

@gadicc: Can't live without it as well, that's why I try to fix stuff although I'm **really** (and I mean **really**) not into writing GNOME extensions :smiley:

nlpsuge commented:
@mendres82 

There is an issue: the References window can't be opened. Please see the error stack trace below:

org.gnome.Shell.Extensions[28431]: Gio.DesktopAppInfo has been moved to a separate platform-specific library. Please update your code to use GioUnix.DesktopAppInfo instead.
                                                             0 get() ["resource:///org/gnome/gjs/modules/core/overrides/Gio.js":524:46]
                                                             1 _init() ["file:///home/homeUser/.local/share/gnome-shell/extensions/another-window-session-manager@gmail.com/prefsCloseWindow.js":742:25]
                                                             2 Row() ["file:///home/homeUser/.local/share/gnome-shell/extensions/another-window-session-manager@gmail.com/prefsCloseWindow.js":296:4]
                                                             3 RuleRow() ["file:///home/homeUser/.local/share/gnome-shell/extensions/another-window-session-manager@gmail.com/prefsCloseWindow.js":366:4]
                                                             4 RuleRowByApp() ["file:///home/homeUser/.local/share/gnome-shell/extensions/another-window-session-manager@gmail.com/prefsCloseWindow.js":738:4]
                                                             5 _sync() ["file:///home/homeUser/.local/share/gnome-shell/extensions/another-window-session-manager@gmail.com/prefsCloseWindow.js":260:40]
                                                             6 _initAppRules() ["file:///home/homeUser/.local/share/gnome-shell/extensions/another-window-session-manager@gmail.com/prefsCloseWindow.js":134:18]
                                                             7 init() ["file:///home/homeUser/.local/share/gnome-shell/extensions/another-window-session-manager@gmail.com/prefsCloseWindow.js":44:18]
                                                             8 fillPreferencesWindow() ["file:///home/homeUser/.local/share/gnome-shell/extensions/another-window-session-manager@gmail.com/prefs.js":30:60]
                                                             9 _loadPrefs() ["resource:///org/gnome/Shell/Extensions/js/extensionPrefsDialog.js":43:24]
                                                             10 AsyncFunctionNext() ["self-hosted":800:27]
                                                             11 _init/GLib.MainLoop.prototype.runAsync/</<() ["resource:///org/gnome/gjs/modules/core/overrides/GLib.js":263:


I'll leave the errors here, in case I don't have time to fix it, as I will be on vacation soon.

mendres82 commented:
@nlpsuge 

I'm looking into it, but unfortunately this one is not so easy to track down.

Testing on fully updated Tumbleweed:

- Changing Gio.DesktopAppInfo to GioUnix.DesktopAppInfo doesn't show the stacktrace anymore, but is still not working.
- Version 50 of the extension also doesn't load preference window.

Testing on a Tumbleweed snapshot from November, 12th 2025:

- Unpatched version 52 **does** load the preference window.

Debug and Verbose mode don't show any helpful hints.

mendres82 commented:
```
gjs: Trying to add GtkBox as a child to an AdwPreferencePage, but only AdwPreferencesPage is allowed
gjs: Trying to add GtkScrolledWindow as a child to an AdwPreferencePage, but only AdwPreferencesPage is allowed
gjs: Trying to add GtkBox as a child to an AdwPreferencePage, but only AdwPreferencesPage is allowed
gjs: Trying to add GtkScrolledWindow as a child to an AdwPreferencePage, but only AdwPreferencesPage is allowed
```

Probably found the reason causing this problem.

mendres82 commented:
@nlpsuge 

A fix is now available in #127.

Tested with latest openSUSE Tumbleweed and Ubuntu 24.04 LTS. Also fixed some other errors.
